### DISCUSSION TITLE
Extracting Entire Flow FIeld

### DISCUSSION THREAD

**Discussion by ZaneFrey** [NatLabRockies/floris#738] [Q&A]
Hi all,

Is there a simple way to extract the entire flow field from a simulation? I am picking up code from a previous user who ran FLORIS in an older version (v3.2 I believe?), and it seems that the simple function of `get_flow_data()` from FlorisInterface no longer exists in that form? This was called after `calculate_wake()`. If there is a new function that does this, is it not in the FlorisInterface framework? My goal is to extract the entire flow field with coordinates, speeds, etc. to ultimately be converted to a netCDF file for later analysis and to be able to run this code in the latest version (v3.5). 

I'm fairly new to the FLORIS framework, so I apologize if this seems like a super basic question! I just haven't been able to find a clear answer in the documentation that works in v3.5. Any help would be greatly appreciated. Thank you!

Z

**Comment by rafmudaf** [comment]
Hi @ZaneFrey,

The `FlorisInterface.get_flow_data()` function was dropped when we moved from v2 to v3. You can see the definition for that function in the [v2.5 tag here](https://github.com/NREL/floris/blob/v2.5/floris/tools/floris_interface.py#L615).

As for doing something similar in the current FLORIS, you could access the `FlowField.u`, `FlowField.v`, and `FlowField.w` variables to get the velocity components at every point in the computational domain for every wind direction and wind speed combination. These variables are 5-dimensional, and I recommend reading the [Background and Concepts](https://nrel.github.io/floris/floris_101.html#build-the-model) page of the documentation to understand each dimension. Also, the coordinates are at `Grid.x`, `Grid.y`, and `Grid.z`.

A small script to get you start is below.

```python
fi = FlorisInterface("inputs/gch.yaml")
fi.calculate_wake()
print(fi.floris.grid.x_sorted)
print(fi.floris.grid.y_sorted)
print(fi.floris.grid.z_sorted)
print(fi.floris.flow_field.u)
print(fi.floris.flow_field.v)
print(fi.floris.flow_field.w)
```

Please let me know if you get stuck. Also, once you integrate this to a netCDF format, please share! I'm interested in this and we could incorporate that into the `floris.tools` package.

**Reply by ZaneFrey** [reply]
Hi, thank you for getting back to me when I initially posted this. I had to step away from this project for a while but am finally getting back to it. I've been digging through your suggestion and the documentation for a bit now and am still trying to make sense of extracting velocities across a domain, not just at the turbine, usually of grid size 3x3. If I want to extract velocities in a wake region or across an entire domain do I have to use a different function through tools, such as something in `flow_data.py`? Or should I go about this differently? Thanks, let me know if this doesn't make too much sense.

**Reply by rafmudaf** [reply]
The way FLORIS handles grid points isn't documented well, so my apologies for that and I'll try to give you guidance here.
For a typical use like an AEP analysis, FLORIS puts grid points on the turbine swept areas as described in [Background and Concepts](https://nrel.github.io/floris/floris_101.html#on-grid-points). Since the wake models are analytical, there's no need to have intermediate points between turbines. For this type of calculation, the quantities I refer to above have the shape `(n wind direction, n wind speed, n turbines, n grid, n grid)` where n grid is usually < 3.

However, for visualization purposes, FLORIS has another type of grid that places points throughout whole wind farm domain in a regular grid. If you're interested, you can see the implementation of this in the [full_flow_sequential_solver](https://github.com/NREL/floris/blob/main/floris/simulation/solver.py#L264) function. There, the turbine-grid points are calculated first in order to obtain the thrust coefficients for all turbines, and then the flow field grid points are calculated given the turbine grid values. In this case, the quantities above have the shape `(1, 1, n x, n y, n z)`. For these calculations, the wind direction and wind speed dimensions are required to be 1 because the data structures get very large. The remaining three dimensions are the number of grid points in x, y, and z.

To switch to a full flow solve, you should set the [solver type](https://github.com/NREL/floris/blob/main/examples/inputs/gch.yaml#L54) to either "flow_field_planar_grid" for 3 planes in the vertical direction or "flow_field_grid" for any number of points in each direction. Note that these two grid types are not very well supported outside of the `FlorisInterface.calculate_xx_plane` functions, so using them directly requires a bit more attention to details. I've outlined the inputs for both below, and you can see the difference in size in the data structures for each.

I hope this gets you started, and please let me know how it goes.

### Planar grid
The "solver" block of your input file would look like blow. Note the two lines after "flow_field_bounds" are intentionally left blank so that they resolve to `None` when the yaml file is loaded.
```yaml
solver:
  type: flow_field_planar_grid
  normal_vector: "z"
  planar_coordinate: 90
  flow_field_grid_points:
  - 200  # number of points in the x direction 
  - 100  # number of points in the y direction
  - 50   # number of points in the z direction
  flow_field_bounds:
  -
  -
```

And a script to use it would be like this:
```python
import numpy as np
from floris.simulation import Floris

f = Floris.from_file("inputs/gch.yaml")
f.farm.yaw_angles = 10 * np.ones([1,1,3])
f.solve_for_viz()
print(np.shape(f.grid.x_sorted))
print(np.shape(f.grid.y_sorted))
print(np.shape(f.grid.z_sorted))
print(np.shape(f.flow_field.u_sorted))
print(np.shape(f.flow_field.v_sorted))
print(np.shape(f.flow_field.w_sorted))
```

Results:
```
$ python 01_opening_floris_computing_power.py 
flow_field.py:130: RuntimeWarning: invalid value encountered in power
  wind_profile_plane = (grid.z_sorted / self.reference_wind_height) ** self.wind_shear
(1, 1, 200, 100, 3)
(1, 1, 200, 100, 3)
(1, 1, 200, 100, 3)
(1, 1, 200, 100, 3)
(1, 1, 200, 100, 3)
(1, 1, 200, 100, 3)
```

### Full flow grid
Inputs:
```yaml
solver:
  type: flow_field_grid
  flow_field_grid_points:
  - 200  # number of points in the x direction 
  - 100  # number of points in the y direction
  - 50   # number of points in the z direction
```

You can use the same script as above, and the results are:
```
$ python 01_opening_floris_computing_power.py 
flow_field.py:130: RuntimeWarning: invalid value encountered in power
  wind_profile_plane = (grid.z_sorted / self.reference_wind_height) ** self.wind_shear
(1, 1, 600, 300, 50)
(1, 1, 600, 300, 50)
(1, 1, 600, 300, 50)
(1, 1, 600, 300, 50)
(1, 1, 600, 300, 50)
(1, 1, 600, 300, 50)
```
